azure/Kqlmagic/adal_token_cache.py

```diff
@@ -117,14 +169,107 @@ def _query_cache(self, is_mrrt, user_id, client_id):
+            user_id = db_key[len("kqlmagicstore/tokens/"):]
+
+            data = token_tuple[1] #encrypted token
+            if not data:
+                return
+            try:
+                time_created, _unused = Fernet._get_unverified_token_data(data)  #get token timestamp 
+            except InvalidToken:
+                Display.showWarningMessage("Warning: found an illegal token, deleting the token")
+                del self.db["kqlmagicstore/tokens/" + user_id ]
+            EXP_TIME = self.token_exp_time  
+            if(time_created +EXP_TIME < int(time.time()) ):
+                del self.db["kqlmagicstore/tokens/" + user_id ]
+
     def _save(self):
         data = self.serialize()
-        self.db['kqlmagicstore/tokens'] = data
+        data_as_bytes = data.encode()
+
+        #init fernet
+        fernet = self.fernet
+
+        #encrypt 
+        data_encrypted = fernet.encrypt(data_as_bytes)
+        
+        self.db['kqlmagicstore/tokens/'+ str(self.user_id)] = data_encrypted
 
     def _restore(self):
+
+        if(AdalTokenCache.LAST_CLEAR_TIME + TIME_BETWEEN_CLEANUPS < int(time.time()) ):
+            AdalTokenCache.LAST_CLEAR_TIME = int(time.time())
+            self.clear_db()
+
+        data_encrypted = self.db.get('kqlmagicstore/tokens/'+ str(self.user_id))
+        if not data_encrypted:
+            return
+        fernet = self.fernet
         try:
-            data = self.db['kqlmagicstore/tokens']
-        except KeyError:
-            print("no stored tokens")
+            data_decrypted_as_bytes = fernet.decrypt(data_encrypted)   
+        except InvalidToken: #Either token has bad form or it cannot be decrypted
+            try: 
+                Fernet._get_unverified_token_data(data_encrypted)
+            except: #the token has bad form
+                Display.showWarningMessage("Warning: found an illegal token, deleting the token")
+                del self.db['kqlmagicstore/tokens/'+ str(self.user_id)]
+                return
+            Display.showWarningMessage("Invalid token, could not activate Single Sign On") #the token cannot be decrypted
+            return
+        data = data_decrypted_as_bytes.decode()
+        self.deserialize(data)
+
+    
+    @staticmethod
+    def get_params_SSO(): #pylint: disable=no-method-argument
+        SSO_id_enc_key = os.getenv("{0}_SSO_ENCRYPTION_KEYS".format(Constants.MAGIC_CLASS_NAME.upper()))
+        key_vals_SSO = Parser.parse_and_get_kv_string(SSO_id_enc_key, {}) if SSO_id_enc_key else {}
+
+        username_SSO = key_vals_SSO.get("username")  
+        secret_key_SSO = key_vals_SSO.get("secretkey")
+        uuid_salt = key_vals_SSO.get("uuid")
+        if uuid_salt and secret_key_SSO:
+            try:
+                uuid_salt = UUID(uuid_salt, version=4)
+            except ValueError:
+                raise ValueError("please enter a valid uuid (version 4) for enabling SSO")
```

Comment:
instead of rasing an exception, just display a message, and add that SSO could not be actiavted